File: ta_service/api/routes/conversations.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status

from ta_service.api.deps import get_conversation_service, get_current_user, get_resolution_service
from ta_service.models.auth import MobileUser
from ta_service.models.conversation import (
    ConversationDetail,
    ConversationSummary,
    CreateConversationRequest,
    PostConversationMessageRequest,
    PostConversationMessageResponse,
)
from ta_service.models.resolution import (
    ResolutionConfirmRequest,
    ResolutionRequest,
    ResolutionResponse,
)
from ta_service.services.conversation_service import ConversationService
from ta_service.services.resolution_service import ResolutionService

router = APIRouter(prefix="/conversations", tags=["conversations"])


@router.post("", response_model=ConversationSummary, status_code=status.HTTP_201_CREATED)
def create_conversation(
    payload: CreateConversationRequest,
    current_user: MobileUser = Depends(get_current_user),
    conversation_service: ConversationService = Depends(get_conversation_service),
) -> ConversationSummary:
    return conversation_service.create_conversation(user_id=current_user.id, title=payload.title)


@router.get("", response_model=list[ConversationSummary])
def list_conversations(
    current_user: MobileUser = Depends(get_current_user),
    conversation_service: ConversationService = Depends(get_conversation_service),
) -> list[ConversationSummary]:
    return conversation_service.list_conversations(user_id=current_user.id)


@router.get("/{conversation_id}", response_model=ConversationDetail)
def get_conversation(
    conversation_id: str,
    current_user: MobileUser = Depends(get_current_user),
    conversation_service: ConversationService = Depends(get_conversation_service),
) -> ConversationDetail:
    conversation = conversation_service.get_conversation(
        user_id=current_user.id,
        conversation_id=conversation_id,
    )
    if not conversation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Conversation not found",
        )
    return conversation


@router.delete("/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_conversation(
    conversation_id: str,
    current_user: MobileUser = Depends(get_current_user),
    conversation_service: ConversationService = Depends(get_conversation_service),
) -> None:
    conversation_service.delete_conversation(
        user_id=current_user.id,
        conversation_id=conversation_id,
    )


@router.post("/{conversation_id}/messages", response_model=PostConversationMessageResponse)
def post_conversation_message(
    conversation_id: str,
    payload: PostConversationMessageRequest,
    current_user: MobileUser = Depends(get_current_user),
    conversation_service: ConversationService = Depends(get_conversation_service),
) -> PostConversationMessageResponse:
    return conversation_service.post_message(
        user_id=current_user.id,
        conversation_id=conversation_id,
        message=payload.message,
    )


@router.post("/{conversation_id}/resolution", response_model=ResolutionResponse)
def resolve_conversation_message(
    conversation_id: str,
    payload: ResolutionRequest,
    current_user: MobileUser = Depends(get_current_user),
    resolution_service: ResolutionService = Depends(get_resolution_service),
) -> ResolutionResponse:
    return resolution_service.resolve_message(
        user_id=current_user.id,
        conversation_id=conversation_id,
        message=payload.message,
    )


@router.post("/{conversation_id}/resolution/confirm", response_model=ResolutionResponse)
def confirm_conversation_resolution(
    conversation_id: str,
    payload: ResolutionConfirmRequest,
    current_user: MobileUser = Depends(get_current_user),
    resolution_service: ResolutionService = Depends(get_resolution_service),
) -> ResolutionResponse:
    return resolution_service.confirm_resolution(
        user_id=current_user.id,
        conversation_id=conversation_id,
        payload=payload,
    )

File: mobile_h5/src/api/conversations.ts
import type {
  ConversationDetail,
  ConversationMessage,
  ConversationSummary,
  CreateConversationRequest,
  PostConversationMessageRequest,
  PostConversationMessageResponse
} from '@/types/conversation'
import type { ResolutionConfirmRequest, ResolutionRequest, ResolutionResponse } from '@/types/resolution'

import { env } from '@utils/env'
import { useAuthStore } from '@stores/auth'
import { request } from './request'

const followupRequestTimeoutMs = 240000

export type StreamMessageEvent =
  | { event: 'started'; userMessage: ConversationMessage }
  | {
      event: 'routing'
      routing_intent?: string | null
      routing_primary_section?: string | null
      routing_fallback_sections?: string[]
      routing_reason?: string | null
      llm_router_ms?: number | null
    }
  | { event: 'delta'; text: string }
  | { event: 'completed'; assistantMessage: ConversationMessage }
  | { event: 'error'; message: string; status_code?: number }

type StreamHandlers = {
  onEvent: (event: StreamMessageEvent) => void
}

export const conversationsApi = {
  createConversation(payload: CreateConversationRequest) {
    return request.post<ConversationSummary>('/conversations', payload).then(response => response.data)
  },
  listConversations() {
    return request.get<ConversationSummary[]>('/conversations').then(response => response.data)
  },
  getConversation(id: string) {
    return request.get<ConversationDetail>(`/conversations/${id}`).then(response => response.data)
  },
  postMessage(conversationId: string, payload: PostConversationMessageRequest) {
    return request
      .post<PostConversationMessageResponse>(`/conversations/${conversationId}/messages`, payload, {
        timeout: followupRequestTimeoutMs
      })
      .then(response => response.data)
  },
  async streamPostMessage(
    conversationId: string,
    payload: PostConversationMessageRequest,
    handlers: StreamHandlers
  ) {
    const authStore = useAuthStore()
    const response = await fetch(`${env.apiBaseUrl}/conversations/${conversationId}/messages/stream`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(authStore.token ? { Authorization: `Bearer ${authStore.token}` } : {}),
      },
      body: JSON.stringify(payload),
    })

    if (!response.ok || !response.body) {
      let message = '流式回复失败，请稍后再试'
      try {
        const data = await response.json()
        if (typeof data?.detail === 'string') {
          message = data.detail
        }
      } catch {
        // ignore JSON parse errors
      }
      throw new Error(message)
    }

    const reader = response.body.getReader()
    const decoder = new TextDecoder('utf-8')
    let buffer = ''

    while (true) {
      const { value, done } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      let boundary = buffer.indexOf('\n\n')
      while (boundary >= 0) {
        const rawEvent = buffer.slice(0, boundary)
        buffer = buffer.slice(boundary + 2)
        const parsed = parseSseEvent(rawEvent)
        if (parsed) {
          handlers.onEvent(parsed)
        }
        boundary = buffer.indexOf('\n\n')
      }
    }
  },
  resolve(conversationId: string, payload: ResolutionRequest) {
    return request
      .post<ResolutionResponse>(`/conversations/${conversationId}/resolution`, payload, {
        timeout: followupRequestTimeoutMs
      })
      .then(response => response.data)
  },
  confirmResolution(conversationId: string, payload: ResolutionConfirmRequest) {
    return request
      .post<ResolutionResponse>(`/conversations/${conversationId}/resolution/confirm`, payload, {
        timeout: followupRequestTimeoutMs
      })
      .then(response => response.data)
  },
  deleteConversation(conversationId: string) {
    return request.delete(`/conversations/${conversationId}`).then(response => response.data)
  }
}

function parseSseEvent(rawEvent: string): StreamMessageEvent | null {
  const lines = rawEvent.split('\n')
  let event = 'message'
  const dataLines: string[] = []

  for (const line of lines) {
    if (line.startsWith('event:')) {
      event = line.slice(6).trim()
      continue
    }
    if (line.startsWith('data:')) {
      dataLines.push(line.slice(5).trim())
    }
  }

  if (!dataLines.length) return null
  try {
    return { event, ...JSON.parse(dataLines.join('\n')) } as StreamMessageEvent
  } catch {
    return null
  }
}
